File: admin_app/lib/features/vehicles/domain/models/logistics_booking.dart
/// Mirrors backend `LogisticsBooking.status` (see `backend/models/LogisticsBooking.js`).
enum LogisticsBookingStatus {
  pending('pending', 'Pending'),
  pendingForDriver('pending_for_driver', 'Awaiting driver'),
  confirmed('confirmed', 'Confirmed'),
  processing('processing', 'Processing'),
  inTransit('in_transit', 'In transit'),
  delivered('delivered', 'Delivered'),
  cancelled('cancelled', 'Cancelled'),
  delayed('delayed', 'Delayed');

  const LogisticsBookingStatus(this.apiValue, this.label);

  final String apiValue;
  final String label;

  static LogisticsBookingStatus fromApi(dynamic raw) {
    if (raw == null) return LogisticsBookingStatus.pending;
    final s = raw.toString().toLowerCase().trim().replaceAll('-', '_');
    for (final v in LogisticsBookingStatus.values) {
      if (v.apiValue == s) return v;
    }
    if (s == 'intransit') return LogisticsBookingStatus.inTransit;
    if (s == 'completed') return LogisticsBookingStatus.delivered;
    if (s == 'ongoing') return LogisticsBookingStatus.inTransit;
    return LogisticsBookingStatus.pending;
  }

  /// Billing, roadmap builder, and dispatch locked once execution has started or finished.
  bool get locksSupervisorEditing =>
      this == LogisticsBookingStatus.processing ||
      this == LogisticsBookingStatus.inTransit ||
      this == LogisticsBookingStatus.delivered ||
      this == LogisticsBookingStatus.cancelled;

  bool get canSearchDriver =>
      this == LogisticsBookingStatus.pending ||
      this == LogisticsBookingStatus.pendingForDriver ||
      this == LogisticsBookingStatus.confirmed;

  /// Supervisor "ACTIVE" tab: orders past initial intake but not yet delivered/cancelled.
  bool get isActivePipeline =>
      this == LogisticsBookingStatus.pendingForDriver ||
      this == LogisticsBookingStatus.confirmed ||
      this == LogisticsBookingStatus.processing ||
      this == LogisticsBookingStatus.inTransit;

  /// Sidebar "Active / Processing" bucket (pre–in-transit work).
  bool get isSidebarPreTransitActive =>
      this == LogisticsBookingStatus.pendingForDriver ||
      this == LogisticsBookingStatus.confirmed ||
      this == LogisticsBookingStatus.processing;
}

class LogisticsSegment {
  final String id;
  final Map<String, dynamic> start;
  final Map<String, dynamic> end;
  final String mode; // Road, Train, Flight, Sea Cargo
  final double distanceKm;
  final String? driverId;
  final String? transportName;
  final String? transportNumber;
  final String? estimatedTime;
  final String? estimatedDate;
  final String status;
  final double price;

  LogisticsSegment({
    required this.id,
    required this.start,
    required this.end,
    required this.mode,
    this.distanceKm = 0.0,
    this.driverId,
    this.transportName,
    this.transportNumber,
    this.estimatedTime,
    this.estimatedDate,
    required this.status,
    this.price = 0.0,
  });

  factory LogisticsSegment.fromJson(Map<String, dynamic> json) {
    return LogisticsSegment(
      id: json['_id'] ?? '',
      start: json['start'] ?? {},
      end: json['end'] ?? {},
      mode: json['mode'] ?? 'Road',
      distanceKm: (json['distanceKm'] as num?)?.toDouble() ?? 0.0,
      driverId: json['driverId'],
      transportName: json['transportName'],
      transportNumber: json['transportNumber'],
      estimatedTime: json['estimatedTime'],
      estimatedDate: json['estimatedDate'],
      status: json['status'] ?? 'pending',
      price: (json['price'] as num?)?.toDouble() ?? 0.0,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'start': start,
      'end': end,
      'mode': mode,
      'distanceKm': distanceKm,
      'driverId': driverId,
      'transportName': transportName,
      'transportNumber': transportNumber,
      'estimatedTime': estimatedTime,
      'estimatedDate': estimatedDate,
      'status': status,
      'price': price,
    };
  }
}

class LogisticsBooking {
  final String id;
  final String userId; // Unique user ID
  final String userName; // Name of user who booked (e.g. Gaurav)
  final String userPhone; // Phone of user who booked
  final String pickupName;
  final String dropName;
  final String modeOfTravel; // Train, Flight, Sea, etc.
  final double price;
  final LogisticsBookingStatus status;
  final DateTime createdAt;
  final double distanceKm;
  final double vehiclePrice;
  final double helperCost;
  final double additionalCharges;
  final double discountAmount;
  final int helperCount;
  final List<dynamic> items;
  final Map<String, dynamic>? pickupAddressDetails;
  final Map<String, dynamic>? receivedAddressDetails;
  final String? railwayStation;
  final String? transportName;
  final String? transportNumber;
  final String? estimatedTime;
  final String? estimatedDate;
  final List<LogisticsSegment> segments;

  LogisticsBooking({
    required this.id,
    required this.userId,
    required this.userName,
    this.userPhone = '',
    required this.pickupName,
    required this.dropName,
    required this.modeOfTravel,
    required this.price,
    required this.status,
    required this.createdAt,
    this.distanceKm = 0.0,
    this.vehiclePrice = 0.0,
    this.helperCost = 0.0,
    this.additionalCharges = 0.0,
    this.discountAmount = 0.0,
    this.helperCount = 0,
    this.items = const [],
    this.pickupAddressDetails,
    this.receivedAddressDetails,
    this.railwayStation,
    this.transportName,
    this.transportNumber,
    this.estimatedTime,
    this.estimatedDate,
    this.segments = const [],
  });

  factory LogisticsBooking.fromJson(Map<String, dynamic> json) {
    // Helper to extract nested address names
    String? getAddressName(dynamic location) {
      if (location == null) return null;
      if (location is String) return location;
      if (location is Map) {
        return location['name'] ??
            location['address'] ??
            location['fullAddress'] ??
            location['label'];
      }
      return null;
    }

    return LogisticsBooking(
      id: json['_id'] ?? '',
      userId: json['userId'] ?? '',
      userName: json['userName'] ?? json['userId'] ?? 'Unknown User',
      userPhone: json['userPhone'] ?? json['mobileNumber'] ?? '',
      pickupName: getAddressName(json['pickup']) ??
          getAddressName(json['pickupAddress']) ??
          json['pickupLocation'] ??
          '',
      dropName: getAddressName(json['dropoff']) ??
          getAddressName(json['receivedAddress']) ??
          json['dropName'] ??
          json['dropoffLocation'] ??
          '',
      modeOfTravel: json['modeOfTravel'] ?? json['vehicleType'] ?? 'Unknown',
      price: (json['totalPrice'] as num?)?.toDouble() ??
          (json['price'] as num?)?.toDouble() ??
          (json['fare'] as num?)?.toDouble() ??
          0.0,
      status: LogisticsBookingStatus.fromApi(json['status']),
      createdAt: DateTime.parse(
          json['createdAt'] ?? DateTime.now().toIso8601String()),
      distanceKm: (json['distanceKm'] as num?)?.toDouble() ?? 0.0,
      vehiclePrice: (json['vehiclePrice'] as num?)?.toDouble() ?? 0.0,
      helperCost: (json['helperCost'] as num?)?.toDouble() ?? 0.0,
      additionalCharges: (json['additionalCharges'] as num?)?.toDouble() ?? 0.0,
      discountAmount: (json['discountAmount'] as num?)?.toDouble() ?? 0.0,
      helperCount: json['helperCount'] ?? 0,
      items: json['items'] ?? [],
      pickupAddressDetails:
          json['pickupAddress'] is Map ? json['pickupAddress'] : null,
      receivedAddressDetails:
          json['receivedAddress'] is Map ? json['receivedAddress'] : null,
      railwayStation: json['railwayStation'],
      transportName: json['transportName'],
      transportNumber: json['transportNumber'],
      estimatedTime: json['estimatedTime']?.toString(),
      estimatedDate: json['estimatedDate']?.toString(),
      segments: (json['segments'] as List?)
              ?.map((s) => LogisticsSegment.fromJson(s))
              .toList() ??
          [],
    );
  }
}

File: backend/models/LogisticsBooking.js
const mongoose = require('mongoose');

// ─── Sub-schemas ────────────────────────────────────────
const locationSchema = new mongoose.Schema({
    name:    { type: String },
    address: { type: String },
    lat:     { type: Number },
    lng:     { type: Number },
}, { _id: false });

const itemSchema = new mongoose.Schema({
    itemName: { type: String, required: true },
    type:     { type: String, default: 'General' }, // Furniture, Electronics, etc.
    length:   { type: Number, default: 0 },
    height:   { type: Number, default: 0 },
    width:    { type: Number, default: 0 },
    unit:     { type: String, default: 'cm' },
    weight:   { type: Number, default: 0 }, // Weight in kg
    quantity: { type: Number, default: 1 }, // Quantity of this item
}, { _id: false });

const addressSchema = new mongoose.Schema({
    type:        { type: String, enum: ['pickup', 'received'], default: null },
    label:       { type: String },
    fullAddress: { type: String },
    houseNumber: { type: String },
    floorNumber: { type: String },
    landmark:    { type: String },
    city:        { type: String },
    pincode:     { type: String },
    phone:       { type: String },
    email:       { type: String },
}, { _id: false });

// ─── Main Logistics Booking Schema ──────────────────────
const logisticsBookingSchema = new mongoose.Schema({

    // Who booked
    userId: {
        type: String,
        required: true,
        index: true,
    },
    userName: {
        type: String,
        default: "",
    },
    userPhone: {
        type: String,
        default: "",
    },

    // Pickup & Drop locations (from map search)
    pickup: { type: locationSchema, required: true },
    dropoff: { type: locationSchema, required: true },

    // Distance & estimated price
    distanceKm:    { type: Number, default: 0 },
    vehiclePrice:  { type: Number, default: 0 },
    helperCost:    { type: Number, default: 0 },
    additionalCharges: { type: Number, default: 0 },
    discountAmount:{ type: Number, default: 0 },
    totalPrice:    { type: Number, default: 0 },
    appliedCoupon: { type: String, default: null },
    
    // Estimates
    estimatedTime: { type: String, default: "" }, // e.g. "2 Days", "10:30 AM"
    estimatedDate: { type: String, default: "" }, // e.g. "2024-03-28"

    // Vehicle
    vehicleType: { type: String, required: true }, // Train, Flight, Sea Cargo, etc.

    // List of items being transported
    items: {
        type: [itemSchema],
        default: [],
    },

    // Helpers
    helperCount: { type: Number, default: 0 },

    // Saved addresses (from address book)
    pickupAddress:   { type: addressSchema, default: null },
    receivedAddress: { type: addressSchema, default: null }, // delivery address

    // Status
    status: {
        type: String,
        enum: ['pending', 'pending_for_driver', 'processing', 'confirmed', 'in_transit', 'delivered', 'cancelled'],
        default: 'pending',
    },

    // Assigned Driver
    driverId: {
        type: mongoose.Schema.Types.ObjectId,
        ref: 'Driver',
        default: null,
    },

    // ─── Multi-Segment Roadmap (Supervisor Controlled) ─────
    segments: [{
        start:           { type: locationSchema },
        end:             { type: locationSchema },
        mode:            { type: String, enum: ['Road', 'Train', 'Flight', 'Sea Cargo'], default: 'Road' },
        distanceKm:      { type: Number, default: 0 },
        driverId:        { type: mongoose.Schema.Types.ObjectId, ref: 'Driver', default: null },
        transportName:   { type: String },   // e.g. Train Name
        transportNumber: { type: String },   // e.g. Train Number
        estimatedTime:   { type: String },   // e.g. "10:30 AM" or "3 Hours"
        estimatedDate:   { type: String },   // e.g. "2024-03-28"
        status: { 
            type: String, 
            enum: ['pending', 'processing', 'completed'], 
            default: 'pending' 
        },
        price:           { type: Number, default: 0 }
      }],

    // Railway Station for Train Mode
    railwayStation: {
        type: String,
        default: null,
    },

    // Transport Details (Dynamic based on mode)
    transportName: {
        type: String,
        default: null,
    },
    transportNumber: {
        type: String,
        default: null,
    },

    // Cancellation charge (if cancelled after free window)
    cancellationCharge: { type: Number, default: 0 },

    // OTP for Hand-off Verification
    otp: {
        type: String,
        default: null,
    },

    // Track drivers who rejected this booking
    rejectedBy: [{
        type: mongoose.Schema.Types.ObjectId,
        ref: 'Driver',
    }],

    // Fare breakdown for transparency
    fareBreakdown: {
        configId: { type: mongoose.Schema.Types.ObjectId, ref: 'PricingConfig' },
        configName: { type: String },
        baseFare: { type: Number, default: 0 },
        distanceCharge: { type: Number, default: 0 },
        weightCharge: { type: Number, default: 0 },
        volumeCharge: { type: Number, default: 0 },
        helperCharge: { type: Number, default: 0 },
        fragileCharge: { type: Number, default: 0 },
        bulkyCharge: { type: Number, default: 0 },
        modeMultiplier: { type: Number, default: 1.0 },
        nightSurcharge: { type: Number, default: 0 },
        tollCharges: { type: Number, default: 0 },
        platformFee: { type: Number, default: 0 },
        gstAmount: { type: Number, default: 0 },
        subtotal: { type: Number, default: 0 },
        totalFare: { type: Number, default: 0 },
    },

}, { timestamps: true });

module.exports = mongoose.model('LogisticsBooking', logisticsBookingSchema);
